File: 2958. Length of Longest Subarray With at Most K Frequency.js
// O(n) time and space complexity
var maxSubarrayLength = function(nums, k) {
    let freq = {}, L = 0, ret = 0;

    nums.map((num, R) => {
        freq[num] = freq[num] ? freq[num] + 1 : 1;

        while (freq[num] > k) {
            freq[nums[L]] = freq[nums[L]] - 1;
            L++;
        };

        ret = ret > R - L + 1 ? ret : R - L + 1;
    });

    return ret;
};

File: 2958. Length of Longest Subarray With at Most K Frequency.py
# O(n) time and space complexity
from collections import defaultdict
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        freq = defaultdict(int)
        L, ret = 0, 0

        for (R, num) in enumerate(nums):
            freq[num] += 1

            while freq[num] > k:
                freq[nums[L]] -= 1
                L += 1

            ret = max(ret, R - L + 1)

        return ret
